src/main/core/work/event.rs

```diff
@@ -61,105 +66,121 @@ impl PartialEq for Event {
         self.magic.debug_check();
         other.magic.debug_check();
 
+        fn check_impl_eq(_: impl Eq) {}
+
+        // below we impl Eq for Event, so we should make sure that all of our comparisons below are
+        // also Eq
+        check_impl_eq(self.time);
+        check_impl_eq(&self.data);
+
         // check every field except '_counter'
-        self.task == other.task
-            && self.time == other.time
-            && self.src_host_id == other.src_host_id
-            && self.dst_host_id == other.dst_host_id
-            && self.src_host_event_id == other.src_host_event_id
+        self.time == other.time && self.data == other.data
     }
 }
 
+// we checked above that Event's `PartialEq` impl is also `Eq`
 impl Eq for Event {}
 
 impl PartialOrd for Event {
     fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
         self.magic.debug_check();
         other.magic.debug_check();
 
-        // sort by event time first, then use other fields we're able to compare
+        // sort by event time, then by the event data
+        let cmp = self.time.cmp(&other.time);
+
+        if cmp.is_ne() {
+            Some(cmp)
+        } else {
+            // event times were equal
+            self.data.partial_cmp(&other.data)
+        }
+    }
+}
+
+/// Data for an event. Different event types will contain different data.
+#[derive(Debug, PartialEq, Eq, PartialOrd)]
+pub enum EventData {
+    // IMPORTANT: The order of these enum variants is important and deliberate. The `PartialOrd`
+    // derive affects the order of events in the event queue, and therefore which events are
+    // processed before others (packet events will be processed before local events), and changing
+    // this could significantly affect the simulation, possibly leading to incorrect behaviour.
+    Packet(PacketEventData),
+    Local(LocalEventData),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct PacketEventData {
+    packet: Packet,
+    src_host_id: HostId,
+    src_host_event_id: u64,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct LocalEventData {
+    task: TaskRef,
+    event_id: u64,
+}
+
+impl PacketEventData {
+    /// Get the packet contained within this event.
+    pub fn packet(self) -> Packet {
+        self.packet
+    }
+}
```

Comment:
Do we want to do this by implementing the `From` trait instead?

Reply:
Sure, I've made it `impl From<PacketEventData> for Packet` instead.